### PR TITLE
feat(data-explorer): map multiple filters by meta fieldtype

### PR DESCRIPTION
**Pull request by MaxPostema**
#### Checklist
- [ ] Functionality works & meets specifications
- [ ] Code reviewed
- [ ] Code unit/integration/system tested
- [ ] User documentation updated
- [ ] Conventional commits (squash if needed)
- [ ] No warnings during install
- [ ] Updated javascript typing

### CODE REVIEW

File: packages/data-explorer/src/mappers/mapperUtils.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    *            text: 'Example option 2'
+    *          }
+    *        ]
+    *    }
+    *  ]
+    * }
+    *
+    * @param attribute
+    * @param options MapperOptions optional object containing options to configure mapper
+    * @returns {Function|null} Function which returns a Promise representing an Array of FieldOptions
```
Comment (connoratrug) [inline packages/data-explorer/src/mappers/mapperUtils.ts]: Suggest to alway return a promise returning function , this makes it easier to handle for caller

File: packages/data-explorer/src/mappers/metaFilterMapper.ts [hunk 2]
```diff
@@ -1,42 +1,77 @@
...
+import { getFieldOptions } from '@/mappers/mapperUtils'
 
-const mapMetaToFilters = async (meta: MetaDataApiResponse) => {
-  let shownFilters:string[] = []
+const MaxVisibleOptions = 10
 
-  // TODO: map all filters
-  const categoricals = await Promise.all(getCategoricals(meta.attributes).map(async (item) => {
-    const href = item && item.refEntity && item.refEntity.href
+const fieldTypeToFilterType:any = {
```
Comment (connoratrug) [inline packages/data-explorer/src/mappers/metaFilterMapper.ts]: can remove the any here, it s a map<string,string) , no typed needed here i think

File: packages/data-explorer/tests/unit/mappers/mapperUtils.spec.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
...bel': 'Gender', 'description': 'gender is located in the root package because it is not hospital specific', 'attributes': [{ 'href': '/api/v2/root_gender/meta/id', 'fieldType': 'STRING', 'name': 'id', 'label': 'id', 'attributes': [], 'maxLength': 255, 'auto': false, 'nillable': false, 'readOnly': true, 'labelAttribute': false, 'unique': true, 'visible': true, 'lookupAttribute': false, 'isAggregatable': false }, { 'href': '/api/v2/root_gender/meta/label', 'fieldType': 'STRING', 'name': 'label', 'label': 'label', 'attributes': [], 'maxLength': 255, 'auto': false, 'nillable': false, 'readOnly': false, 'labelAttribute': true, 'unique': false, 'visible': true, 'lookupAttribute': true, 'isAggregatable': false }], 'labelAttribute': 'label', 'idAttribute': 'id', 'lookupAttributes': ['label'], 'isAbstract': false, 'writable': true, 'languageCode': 'en', 'permissions': ['UPDATE_DATA', 'DELETE_METADATA', 'AGGREGATE_DATA', 'ADD_DATA', 'READ_DATA', 'UPDATE_METADATA', 'READ_METADATA', 'COUNT_DATA', 'DELETE_DATA'] }, 'start': 0, 'num': 100, 'total': 2, 'items': [{ '_href': '/api/v2/root_gender/f', 'id': 'f', 'label': 'Female' }, { '_href': '/api/v2/root_gender/m', 'id': 'm', 'label': 'Male' }] })
```
Comment (connoratrug) [inline packages/data-explorer/tests/unit/mappers/mapperUtils.spec.ts]: very long string makes test hard to maintain, only mock values needed, or if you really need a large data structure ( usually not the case of a unit test) move it to a human readable data structure ( js object)

File: packages/data-explorer/tests/unit/mappers/mapperUtils.spec.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      const result = generateUri(uri, id, label, search)
+      expect(result).toEqual(uri)
+    })
+
+  it('should generate an uri with like query when there is a search string',
+    () => {
+      const search = 'blaat'
+      const result = generateUri(uri, id, label, search)
+      const expected = `${uri}?q=${id}=like=${search},${label}=like=${search}`
+      expect(result).toEqual(expected)
```
Comment (connoratrug) [inline packages/data-explorer/tests/unit/mappers/mapperUtils.spec.ts]: suggest writing out the expected string result, this makes it easier to follow along and spot mistakes in the test

File: packages/data-explorer/tests/unit/mappers/mapperUtils.spec.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  it('should generate an uri without query when there is an empty search array',
+    () => {
+      const search:string[] = []
+      const result = generateUri(uri, id, label, search)
+      expect(result).toEqual(uri)
+    })
+})
+
+describe('getFieldOptions', () => {
+  it('-',
```
Comment (connoratrug) [inline packages/data-explorer/tests/unit/mappers/mapperUtils.spec.ts]: please a test label

File: packages/data-explorer/tests/unit/mappers/mapperUtils.spec.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    (done) => {
+      // Canonical
```
Comment (connoratrug) [inline packages/data-explorer/tests/unit/mappers/mapperUtils.spec.ts]: suggest moving to separate 'it's' and removing the conditional expect checks

**Comment by connoratrug** [comment]
all unit test succeed but output still shows some difference between expected and received

```
Time:        10s(node:41895) UnhandledPromiseRejectionWarning: Error: expect(received).toEqual(expected)

Expected value to equal:
  [{"text": "Female", "value": "f"}, {"text": "Male", "value": "m"}]
Received:
  [{"text": "chair", "value": "chair"}, {"text": "table", "value": "table"}]

Difference:

- Expected
+ Received

  Array [
    Object {
-     "text": "Female",
-     "value": "f",
+     "text": "chair",
+     "value": "chair",
    },
    Object {
-     "text": "Male",
-     "value": "m",
+     "text": "table",
```

File: packages/data-explorer/package.json [hunk 7]
```diff
@@ -24,7 +24,7 @@
     "@molgenis/molgenis-api-client": "^3.1.3",
     "@molgenis/molgenis-i18n-js": "^0.2.1",
     "@molgenis/molgenis-ui-context": "^1.3.4",
-    "@molgenis/molgenis-ui-filter": "^1.1.2",
+    "@molgenis/molgenis-ui-filter": "^1.4.0",
```
Comment (sidohaakma) [inline packages/data-explorer/package.json]: This is now 1.4.2 and how does this relate to this PR: https://github.com/molgenis/molgenis-frontend/pull/173

molgenis-ui-filter@1.4.2 contains: "activeFilter wil not brake when updated quickly"

File: packages/data-explorer/src/mappers/mapperUtils.ts [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    })
+  })
+}
+
+const generateUri = (
+  uri: string, idAttribute: string, labelAttribute: string, search?: string | Array<string> | null) => {
+  if (search) {
+    if (Array.isArray(search) && search.length > 0) {
+      // Join array into a string
+      const value = search.join(',')
```
Comment (fdlk) [inline packages/data-explorer/src/mappers/mapperUtils.ts]: use rsql library to properly escape the search strings

File: packages/data-explorer/src/mappers/mapperUtils.ts [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+      // Use =in= query
+      uri = `${uri}?q=${idAttribute}=in=(${value}),${labelAttribute}=in=(${value})`
```
Comment (fdlk) [inline packages/data-explorer/src/mappers/mapperUtils.ts]: Is two different cases: if you're typing in a filter MREF you want to a search query in the lookup attributes. If you're retrieving the labels of a selected range of IDs you want to do an in query on the id attributes.

File: packages/data-explorer/package.json (hunk 7 above)
Comment (MaxPostema) [reply packages/data-explorer/package.json]: updated

**Pull request closed without merging**